### ISSUE TITLE
Switch to CIM cmdlets

### ISSUE THREAD

**Issue by exchange12rocks** [bcwilhite/PendingReboot#8]
Hi Brian! As far as I know, WMI cmdlets are considered deprecated. Do you think it's worth to switch to CIM cmdlets? Or implement both and let users to choose which one to use via a switch parameter?

**Comment by bcwilhite** [comment]
Actually this was on my to-do list at one point.  I think it's a good idea to switch over given that v3 should be the least version widely used.  I suppose a switch parameter could be used in order to leverage WMI, although I'd just prefer to go one way or the other and not have to support both.

**Comment by alayac** [comment]
This is a great module and we use it extensively in our organization (along with it's predecessor from Technet Gallery).
I just tried this module in PowerShell 7 and it doesn't like the WMI cmdlets.  Migrating to the CIM cmdlets should make this work in the latest version of PowerShell natively.  No need to use a switch operator to support both.  The CIM cmdlets are comparable and better optimized anyway.
Thanks!

**Comment by dennisl68-castra** [comment]
Since PoSH 7.1 is 99% compatbile with PoSH 5.1, it's a shame that the WMI-dependency still exists in this module...
Why not just drop the WMI-support and move to a v2 release?

**Comment by realslacker** [comment]
@bcwilhite if you merge this and someone needs WMI support they can just install the last release before switching to CIM. The current version has been working for people for years, so they should be able to just keep using it. Maybe just putting some explicit install instructions so people understand that to use WMI they use 0.9.0.6, anything after that uses CIM.

**Comment by DennisL68** [comment]
It seems that CIM itself is supported in PowerShell since v3...

I.e. natively in Windows 8 and later. 

So who is actually bound to using WMI if not using an unsupported Windows version?

https://devblogs.microsoft.com/scripting/what-is-cim-and-why-should-i-use-it-in-powershell/

https://devblogs.microsoft.com/powershell/introduction-to-cim-cmdlets/

**Comment by realslacker** [comment]
@DennisL68 CIM uses WinRM (WSMAN) where WMI uses native WMI calls. I've come across a few scenarios where WMI is enabled, but WinRM is not, and vice-versa.

**Comment by DennisL68** [comment]
> @DennisL68 CIM uses WinRM (WSMAN) where WMI uses native WMI calls. I've come across a few scenarios where WMI is enabled, but WinRM is not, and vice-versa.

Yes, WMI uses RPC where as CIM uses a more controlled protocol as a carrier :)

**Comment by SamErde** [comment]
It shouldn't be hard to include an option to use WMI as a fallback method.

**Comment by ShaneJ7769** [comment]
Why not just do a regkey check?

# Check for pending reboot that might be required after a domain join
if (Test-Path -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\RebootRequired") {
    Write-Output "Pending reboot detected. Domain join might not be complete."
}

# Check for pending computer rename
if (Test-Path -Path "HKLM:\SYSTEM\CurrentControlSet\Control\ComputerName\ActiveComputerName") {
    $newName = (Get-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\ComputerName\ComputerName").ComputerName
    $currentName = $env:COMPUTERNAME
    if ($newName -ne $currentName) {
        Write-Output "Pending computer rename detected. Domain join might not be complete."
    }
}

**Comment by ShaneJ7769** [comment]
Those looking for a work around:

$RebootFlags = Test-PendingReboot -Detailed -WarningAction SilentlyContinue -ErrorAction SilentlyContinue
$JoinFlag = (Test-Path -Path "HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\RebootRequired")
$PendingRename = ((Test-Path -Path "HKLM:\SYSTEM\CurrentControlSet\Control\ComputerName\ActiveComputerName") -and ((Get-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\ComputerName\ComputerName").ComputerName -ne $env:COMPUTERNAME))

if ($RebootFlags.IsRebootPending -eq $true) { 
   if (($RebootFlags.ComponentBasedServicing -eq $True) -or `
       #($RebootFlags.PendingComputerRenameDomainJoin -eq $True) -or `
       ($RebootFlags.PendingFileRenameOperations -eq $True) -or `
       ($RebootFlags.SystemCenterConfigManager -eq $True) -or `
       ($RebootFlags.WindowsUpdateAutoUpdate -eq $True) -or `
       ($JoinFlag -eq $true) -or `
       ($PendingRename -eq $True)) {
	  Write-Host "There is a pending reboot needed." -ForegroundColor Red 
}}

**Comment by DennisL68** [comment]
So, have many years have this been an issue now?

**Comment by exchange12rocks** [comment]
@DennisL68 Fell free to try this one: https://github.com/exchange12rocks/PendingRebootCIM
